### DISCUSSION TITLE
Project Settings

### DISCUSSION THREAD

**Discussion by Martyx00** [Vector35/binaryninja-api#2120] [API] [answered]
Hello,
I wanted to ask if someone has a working example on how to initialize and use settings with `Project` scope only. I do know that we have this page: https://api.binary.ninja/binaryninja.settings-module.html but since I am not the brightest developer I was not able to figure out how to create a project-only setting of type JSON that will be stored in BNDB file (so that it can be shared in case of collaboration).

Thanks in adavance!

**Comment by psifertex** [comment, marked as answer]
Depending on what you want to use it for, the [store_metadata](https://api.binary.ninja/binaryninja.binaryview-module.html#binaryninja.binaryview.BinaryView.store_metadata) / [query_metadata](https://api.binary.ninja/binaryninja.binaryview-module.html#binaryninja.binaryview.BinaryView.query_metadata) APIs are a bit simpler to use.

That said, here's a quick example that uses the resource-based settings. Note that a project-based setting will actually create a `.binaryninja/` folder and put the settings in there for multiple files. I don't think you want that. A resource setting will put the setting in the BNDB itself:

```
>>> import binaryninja
>>> bv = binaryninja.open_view('/tmp/ls')
>>> plugin_settings = binaryninja.Settings()
>>> plugin_settings.register_setting("ui.dothething", '{"description" : "Tells Zhu Li to do the thing.", "title" : "Do the thing", "default" : true, "type" : "boolean"}')
True
>>> plugin_settings.set_bool('ui.dothething', False, bv, binaryninja.SettingsScope.SettingsResourceScope)
True
>>> #Resource scope will save the setting in the BNDB itself
>>> bv.create_database('/tmp/ls.bndb')
True
>>> bv = binaryninja.open_view('/tmp/ls')
>>> #Dont have to register the setting since it is already registered in this settings instance, otherwise we would
>>> plugin_settings.get_bool('ui.dothething', bv)
True
>>> bv = binaryninja.open_view('/tmp/ls.bndb')
>>> plugin_settings.get_bool('ui.dothething', bv)
False
```

As long as your plugin always registers that schema users will be able to use the settings UI to set/clear the settings value. Obviously you can use the metadata API if you don't want to use the settings API. 

Note that on the query side, you can just `get_bool` and the settings system itself will walk the hierarchy of different settings to find the most specific one (as described in the docs you linked, Resource is always the most-specific only overridden by settings the user themselves set using "open with options")

Thanks for the good question, I'll go ahead and add this example to the API docs as well.

**Reply by psifertex** [reply]
Updated on dev, will be live online in the next stable: https://github.com/Vector35/binaryninja-api/blob/dev/python/settings.py#L64
